# Feature selection

## 0 Libraries, functions

In [ ]:
# ! pip install feature-engine
# ! pip install pycaret
! pip install xgboost


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split,cross_validate

## Constant and Quasi-constant features with Feature-engine
from feature_engine.selection import (
    DropConstantFeatures,  DropDuplicateFeatures,
SmartCorrelatedSelection,)

# Correlation
from feature_engine.selection.smart_correlation_selection import SmartCorrelatedSelection

# Pipleline
from sklearn.pipeline import Pipeline

# For select variables
from sklearn.feature_selection import (SelectFromModel,  RFE)

# For scan models
from pycaret.classification import *

# Models
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

#Scale
from sklearn.preprocessing import MinMaxScaler

from xgboost import XGBClassifier


### 0.2 Load dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# load dataset baseline
data_dir = ('/content/drive/MyDrive/Colab Notebooks/tfm/data_fe/')
X_train = pd.read_csv(data_dir + 'X_train_bl.csv')
X_test = pd.read_csv(data_dir + 'X_test_bl.csv')
y_train = pd.read_csv(data_dir + 'y_train_bl.csv')
y_test = pd.read_csv(data_dir + 'y_test_bl.csv')

In [ ]:
print(X_train.shape,
y_train.shape, X_test.shape, y_test.shape)

(320978, 18) (319879, 1) (69183, 18) (69183, 1)


In [ ]:
# load dataset 2
data_dir = ('/content/drive/MyDrive/Colab Notebooks/tfm/data_fe/')
X_train_2 = pd.read_csv(data_dir + 'X_train_2.csv')
X_test_2 = pd.read_csv(data_dir + 'X_test_2.csv')
y_train_2 = pd.read_csv(data_dir + 'y_train_2.csv')
y_test_2 = pd.read_csv(data_dir + 'y_test_2.csv')

In [ ]:
print(X_train_2.shape,
y_train_2.shape)

(320978, 25) (320978, 1)


In [ ]:
# load dataset 2a
data_dir = ('/content/drive/MyDrive/Colab Notebooks/tfm/data_fe/')
X_train_2a = pd.read_csv(data_dir + 'X_train_2a.csv')
X_test_2a = pd.read_csv(data_dir + 'X_test_2a.csv')
y_train_2a = pd.read_csv(data_dir + 'y_train_2a.csv')
y_test_2a = pd.read_csv(data_dir + 'y_test_2a.csv')

In [ ]:
print(X_train_2a.shape,
y_train_2a.shape)

(322474, 38) (322474, 1)


In [ ]:
# load dataset 3
data_dir = ('/content/drive/MyDrive/Colab Notebooks/tfm/data_fe/')
X_train_3 = pd.read_csv(data_dir + 'X_train_3.csv')
X_test_3 = pd.read_csv(data_dir + 'X_test_3.csv')
y_train_3 = pd.read_csv(data_dir + 'y_train_3.csv')
y_test_3 = pd.read_csv(data_dir + 'y_test_3.csv')

In [ ]:
print(X_train_3.shape,
y_train_3.shape)

(314327, 66) (314327, 1)


In [ ]:
# load dataset baseline
data_dir = ('/Users/aprilm/Desktop/TFG/data_fe/')
X_train = pd.read_csv(data_dir + 'X_train_bl.csv')
X_test = pd.read_csv(data_dir + 'X_test_bl.csv')
y_train = pd.read_csv(data_dir + 'y_train_bl.csv')
y_test = pd.read_csv(data_dir + 'y_test_bl.csv')

In [ ]:
X_train.shape

(319879, 17)

In [ ]:
# load dataset 2
data_dir = ('/Users/aprilm/Desktop/TFG/data_fe/')
X_train_2 = pd.read_csv(data_dir + 'X_train_2.csv')
X_test_2 = pd.read_csv(data_dir + 'X_test_2.csv')
y_train_2 = pd.read_csv(data_dir + 'y_train_2.csv')
y_test_2 = pd.read_csv(data_dir + 'y_test_2.csv')

In [ ]:
X_train_2.shape

(320978, 25)

In [ ]:
# load dataset 2a
data_dir = ('/Users/aprilm/Desktop/TFG/data_fe/')
X_train_2a = pd.read_csv(data_dir + 'X_train_2a.csv')
X_test_2a = pd.read_csv(data_dir + 'X_test_2a.csv')
y_train_2a = pd.read_csv(data_dir + 'y_train_2a.csv')
y_test_2a = pd.read_csv(data_dir + 'y_test_2a.csv')

In [ ]:
X_train_2a.shape

(322474, 38)

In [ ]:
# load dataset 3
data_dir = ('/Users/aprilm/Desktop/TFG/data_fe/')
X_train_3 = pd.read_csv(data_dir + 'X_train_3.csv')
X_test_3 = pd.read_csv(data_dir + 'X_test_3.csv')
y_train_3 = pd.read_csv(data_dir + 'y_train_3.csv')
y_test_3 = pd.read_csv(data_dir + 'y_test_3.csv')

In [ ]:
X_train_3.shape

(314327, 66)

## 0.3 Dataset copy

In [ ]:
#Baseline
X_train_c = X_train.copy()
X_test_c =  X_test.copy()
y_train_c = y_train.copy()
y_test_c = y_test.copy()

#Dataset 2
X_train_2c = X_train_2.copy()
X_test_2c =  X_test_2.copy()
y_train_2c = y_train_2.copy()
y_test_2c = y_test_2.copy()

#Dataset 2a
X_train_2ac = X_train_2a.copy()
X_test_2ac =  X_test_2a.copy()
y_train_2ac = y_train_2a.copy()
y_test_2ac = y_test_2a.copy()


#Dataset 3
X_train_3c = X_train_3.copy()
X_test_3c =  X_test_3.copy()
y_train_3c = y_train_3.copy()
y_test_3c = y_test_3.copy()

## 1 Filters methods

### 1.1 Filter methods - constant, quasi constant and duplicated features

##### 1.1.1 Constant features

In [ ]:
# To remove constant features we will use Feature-engine
sel = DropConstantFeatures(tol=1, variables=None, missing_values='raise')

sel.fit(X_train_c)

DropConstantFeatures()

In [ ]:
# number of constant features

len(sel.features_to_drop_)

0

##### 1.1.2 Quasi constant features

In [ ]:
# To remove quai constant features we will use Feature-engine
sel = DropConstantFeatures(tol=0.998, variables=None, missing_values='raise')
# we set the tolerance in 0.998

sel.fit(X_train_c)

DropConstantFeatures(tol=0.998)

In [ ]:
# number of quasi constant features

len(sel.features_to_drop_)

0

##### 1.1.2 Duplicate features

In [ ]:
# To remove duplicate features we will use Feature-engine
# set up the selector
sel = DropDuplicateFeatures(variables=None, missing_values='raise')

# find the duplicate features, this might take a while
sel.fit(X_train_c)

DropDuplicateFeatures(missing_values='raise')

In [ ]:
len(sel.features_to_drop_)

0

##### 1.1.3 Correlation SmartCorrelationSelection - Model perfomance

We will keep a feature from each correlation group based on the performance of a random forest.

We considered to Spearman's Rank. We think that Spearman will be more efficient because can handle both continuous and ordinal varibles. Both types of variables are in our dataset.

In [ ]:
# random forest
rf = RandomForestClassifier(
    n_estimators=10,
    random_state=0,
    n_jobs=4,
)

# correlation selector
sel = SmartCorrelatedSelection(
    variables=None, # if none, selector examines all numerical variables
    method="spearman",
    threshold=0.8,
    missing_values="raise",
    selection_method="model_performance",
    estimator=rf,
    scoring="f1_macro",
    cv=3,
)

# this may take a while, because we are training
# a random forest per correlation group

sel.fit(X_train_c, y_train_c)

SmartCorrelatedSelection(estimator=RandomForestClassifier(n_estimators=10,
                                                          n_jobs=4,
                                                          random_state=0),
                         method='spearman', missing_values='raise',
                         scoring='f1_macro',
                         selection_method='model_performance')

In [ ]:
sel.correlated_feature_sets_

[]

#### 1.1.4 Correlation SmartCorrelationSelection - Variance

In [ ]:
sel = SmartCorrelatedSelection(
    variables=None,
    method="spearman",
    threshold=0.8,
    missing_values="raise",
    selection_method="variance",
    estimator=None,
    scoring="f1_macro",
    cv=3,
)

sel.fit(X_train_c, y_train_c)

SmartCorrelatedSelection(method='spearman', missing_values='raise',
                         scoring='f1_macro', selection_method='variance')

In [ ]:
sel.correlated_feature_sets_

[]

### 1.2 Pipeline filter methods

#### 1.2.1 Pipeline feature selection

In [ ]:
# Define your random forest estimator
rf = RandomForestClassifier(
    n_estimators=10,
    random_state=0,
    n_jobs=4,
)
# we stack all the selection methods inside a pipeline
pipe_model_performance = Pipeline([
    ('constant', DropConstantFeatures(tol=0.998)),
    ('duplicated', DropDuplicateFeatures()),
    ('correlation_model_performance', SmartCorrelatedSelection(
        method='spearman',
        missing_values='raise',
        scoring='f1_macro',
        selection_method='model_performance',
        estimator=rf,
        cv=3,
    )),
])
pipe_model_performance.fit(X_train_c, y_train_c)


Pipeline(steps=[('constant', DropConstantFeatures(tol=0.998)),
                ('duplicated', DropDuplicateFeatures()),
                ('correlation_model_performance',
                 SmartCorrelatedSelection(estimator=RandomForestClassifier(n_estimators=10,
                                                                           n_jobs=4,
                                                                           random_state=0),
                                          method='spearman',
                                          missing_values='raise',
                                          scoring='f1_macro',
                                          selection_method='model_performance'))])

#### 1.2.2 Pipeline variance

In [ ]:
# we stack all the selection methods inside a pipeline

pipe_variance = Pipeline([
    ('constant', DropConstantFeatures(tol=0.998)),
    ('duplicated', DropDuplicateFeatures()),
    ('correlation_variance', SmartCorrelatedSelection(
        method='spearman',
        missing_values='raise',
        scoring='f1_macro',
        selection_method='variance'
    )),
])
pipe_variance.fit(X_train_c)

Pipeline(steps=[('constant', DropConstantFeatures(tol=0.998)),
                ('duplicated', DropDuplicateFeatures()),
                ('correlation_variance',
                 SmartCorrelatedSelection(method='spearman',
                                          missing_values='raise',
                                          scoring='f1_macro',
                                          selection_method='variance'))])

In [ ]:
# X_train
X_train_performance = pipe_model_performance.transform(X_train_c)
X_train_variance = pipe_variance.transform(X_train_c)

#X_test
X_test_performance = pipe_model_performance.transform(X_test_c)
X_test_variance = pipe_variance.transform(X_test_c)

print('X_train_c', X_train_performance.shape, X_train_variance.shape, '\n',
      'X_test_c', X_test_performance.shape, X_test_variance.shape )



X_train_c (319879, 17) (319879, 17) 
 X_test_c (69183, 17) (69183, 17)


## 2 Select features with tree importance

### 2.1 RandomForestClassifier

In [ ]:
# SelectFrom model will select those features which importance
# is greater than the mean importance of all the features
# by default, but you can alter this threshold if you want to

sel_ = SelectFromModel(RandomForestClassifier(n_estimators=10, random_state=0))
sel_.fit(X_train_c, y_train_c)

SelectFromModel(estimator=RandomForestClassifier(n_estimators=10,
                                                 random_state=0))

In [ ]:
sel_.get_support()

array([ True,  True,  True, False, False,  True,  True,  True, False,
        True, False, False, False, False, False, False, False])

In [ ]:
selected_feat = X_train.columns[(sel_.get_support())]

len(selected_feat)

7

In [ ]:
selected_feat

Index(['distance', 'density', 'age', 'diversity', 'green', 'temp', 'wind'], dtype='object')

#### 2.1.1 RandomForestClassifier + filter varianza

In [ ]:
# usign the X_train_variance
sel_.fit(X_train_variance, y_train_c)

SelectFromModel(estimator=RandomForestClassifier(n_estimators=10,
                                                 random_state=0))

In [ ]:
selected_feat_variance = X_train_variance.columns[(sel_.get_support())]

len(selected_feat_variance)

7

In [ ]:
selected_feat_variance

Index(['distance', 'density', 'age', 'diversity', 'green', 'temp', 'wind'], dtype='object')

## 2.2 Pipeline with tree importance

##### 2.2.1 Pipeline only tree importance

In [ ]:
# Define your random forest estimator
rf = RandomForestClassifier(
    n_estimators=10,
    random_state=0,
    n_jobs=4,
)
# Define your random forest model
rf_selector = RandomForestClassifier(n_estimators=10, random_state=0)
sel_ = SelectFromModel(rf_selector)

In [ ]:
# we stack all the selection methods inside a pipeline
pipe_tree_importance = Pipeline([
    ('constant', DropConstantFeatures(tol=0.998)),
    ('duplicated', DropDuplicateFeatures()),
    ('tree_importance', SelectFromModel(rf_selector))

])
pipe_tree_importance.fit(X_train_c, y_train_c)

Pipeline(steps=[('constant', DropConstantFeatures(tol=0.998)),
                ('duplicated', DropDuplicateFeatures()),
                ('tree_importance',
                 SelectFromModel(estimator=RandomForestClassifier(n_estimators=10,
                                                                  random_state=0)))])

#### 2.2.2 Pipeline Correlation model perfomance  + tree importance

In [ ]:
# Define your random forest estimator
rf = RandomForestClassifier(
    n_estimators=10,
    random_state=0,
    n_jobs=4,
)
# Define your random forest model
rf_selector = RandomForestClassifier(n_estimators=10, random_state=0)
sel_ = SelectFromModel(rf_selector)

# we stack all the selection methods inside a pipeline
pipe_model_performance_tree = Pipeline([
    ('constant', DropConstantFeatures(tol=0.998)),
    ('duplicated', DropDuplicateFeatures()),
    ('correlation_model_performance', SmartCorrelatedSelection(
        method='spearman',
        missing_values='raise',
        scoring='f1_macro',
        selection_method='model_performance',
        estimator=rf,
        cv=3,
    )),
    ('tree_importance', SelectFromModel(rf_selector))

])
pipe_model_performance_tree.fit(X_train_c, y_train_c)

Pipeline(steps=[('constant', DropConstantFeatures(tol=0.998)),
                ('duplicated', DropDuplicateFeatures()),
                ('correlation_model_performance',
                 SmartCorrelatedSelection(estimator=RandomForestClassifier(n_estimators=10,
                                                                           n_jobs=4,
                                                                           random_state=0),
                                          method='spearman',
                                          missing_values='raise',
                                          scoring='f1_macro',
                                          selection_method='model_performance')),
                ('tree_importance',
                 SelectFromModel(estimator=RandomForestClassifier(n_estimators=10,
                                                                  random_state=0)))])

#### 2.2.3 Pipeline Correlation variance  + tree importance

In [ ]:
# we stack all the selection methods inside a pipeline
pipe_model_variance_tree = Pipeline([
    ('constant', DropConstantFeatures(tol=0.998)),
    ('duplicated', DropDuplicateFeatures()),
    ('correlation_model_performance', SmartCorrelatedSelection(
        method='spearman',
        missing_values='raise',
        scoring='f1_macro',
        selection_method='variance',
        estimator=rf,
        cv=3,
    )),
    ('tree_importance', SelectFromModel(rf_selector))

])
pipe_model_variance_tree.fit(X_train_c, y_train_c)

Pipeline(steps=[('constant', DropConstantFeatures(tol=0.998)),
                ('duplicated', DropDuplicateFeatures()),
                ('correlation_model_performance',
                 SmartCorrelatedSelection(estimator=RandomForestClassifier(n_estimators=10,
                                                                           n_jobs=4,
                                                                           random_state=0),
                                          method='spearman',
                                          missing_values='raise',
                                          scoring='f1_macro',
                                          selection_method='variance')),
                ('tree_importance',
                 SelectFromModel(estimator=RandomForestClassifier(n_estimators=10,
                                                                  random_state=0)))])

## 3 Pipelines applied at the databases

### 3.1 Pipelines Baseline

In [ ]:
# X_train
X_train_tree = pipe_tree_importance.transform(X_train_c)
X_train_performance_tree = pipe_model_performance_tree.transform(X_train_c)
X_train_variance_tree = pipe_model_variance_tree.transform(X_train_c)

# X_test
X_test_tree = pipe_tree_importance.transform(X_test_c)
X_test_performance_tree = pipe_model_performance_tree.transform(X_test_c)
X_test_variance_tree = pipe_model_variance_tree.transform(X_test_c)

print('X_train', X_train_tree.shape, X_train_performance_tree.shape, X_train_variance_tree.shape, '\n',
      'X_test', X_test_tree.shape, X_test_performance_tree.shape,X_test_variance_tree.shape )

X_train (319879, 7) (319879, 7) (319879, 7) 
 X_test (69183, 7) (69183, 7) (69183, 7)


### 3.2 Pipelines Dataset 2

In [ ]:
# Fitting the pipelines with the training data
pipe_model_performance.fit(X_train_2c, y_train_2c)
pipe_variance.fit(X_train_2c)
pipe_tree_importance.fit(X_train_2c, y_train_2c)
pipe_model_performance_tree.fit(X_train_2c, y_train_2c)
pipe_model_variance_tree.fit(X_train_2c, y_train_2c)

# Transforming X_train with the pipelines
X_train_pefomance_2_pl = pipe_model_performance.transform(X_train_2c)
X_train_variance2_pl = pipe_variance.transform(X_train_2c)
X_train_tree2_pl = pipe_tree_importance.transform(X_train_2c)
X_train_performance_tree2_pl = pipe_model_performance_tree.transform(X_train_2c)
X_train_variance_tree2_pl = pipe_model_variance_tree.transform(X_train_2c)

# Transforming X_test with the pipelines
X_test_pefomance_2_pl = pipe_model_performance.transform(X_test_2c)
X_test_variance2_pl = pipe_variance.transform(X_test_2c)
X_test_tree2_pl = pipe_tree_importance.transform(X_test_2c)
X_test_performance_tree2_pl = pipe_model_performance_tree.transform(X_test_2c)
X_test_variance_tree2_pl = pipe_model_variance_tree.transform(X_test_2c)


In [ ]:
print('X_train', X_train_pefomance_2_pl .shape, X_train_variance2_pl.shape, X_train_tree2_pl.shape, X_train_performance_tree2_pl.shape,X_train_variance_tree2_pl.shape,  '\n',
      'X_test',X_test_pefomance_2_pl.shape, X_test_variance2_pl.shape,X_test_tree2_pl.shape, X_test_performance_tree2_pl.shape, X_test_variance_tree2_pl.shape )

X_train (320978, 18) (320978, 18) (320978, 7) (320978, 7) (320978, 7) 
 X_test (69183, 18) (69183, 18) (69183, 7) (69183, 7) (69183, 7)


### 3.2 Pipelines Dataset 2a

In [ ]:
# Fitting the pipelines with the new data

pipe_model_performance.fit(X_train_2ac, y_train_2ac) #correlation model perfomance
pipe_variance.fit(X_train_2ac) #correlation variance
pipe_tree_importance.fit(X_train_2ac,y_train_2ac) #tree importance
pipe_model_performance_tree.fit(X_train_2ac, y_train_2ac)  #tree + model perfomance
pipe_model_variance_tree.fit(X_train_2ac, y_train_2ac) #tree + variance

#X_train
X_train_pefomance_2a_pl = pipe_model_performance.transform(X_train_2ac)
X_train_variance2a_pl= pipe_variance.transform(X_train_2ac)
X_train_tree2a_pl= pipe_tree_importance.transform(X_train_2ac)
X_train_performance_tree2a_pl= pipe_model_performance_tree.transform(X_train_2ac)
X_train_variance_tree2a_pl= pipe_model_variance_tree.transform(X_train_2ac)


# Transforming X_test with the pipelines (No fitting on the test set)
X_test_pefomance_2a_pl = pipe_model_performance.transform(X_test_2ac)
X_test_variance2a_pl = pipe_variance.transform(X_test_2ac)
X_test_tree2a_pl = pipe_tree_importance.transform(X_test_2ac)
X_test_performance_tree2a_pl = pipe_model_performance_tree.transform(X_test_2ac)
X_test_variance_tree2a_pl = pipe_model_variance_tree.transform(X_test_2ac)


In [ ]:
print('X_train', X_train_pefomance_2a_pl .shape, X_train_variance2a_pl.shape, X_train_tree2a_pl.shape, X_train_performance_tree2a_pl.shape,X_train_variance_tree2a_pl.shape,  '\n',
      'X_test',X_test_pefomance_2a_pl.shape, X_test_variance2a_pl.shape,X_test_tree2a_pl.shape, X_test_performance_tree2a_pl.shape, X_test_variance_tree2a_pl.shape )

X_train (322474, 19) (322474, 19) (322474, 18) (322474, 8) (322474, 7) 
 X_test (69183, 19) (69183, 19) (69183, 18) (69183, 8) (69183, 7)


### 3.3 Pipelines Dataset 3

In [ ]:
# Fitting the pipelines with the new data

pipe_model_performance.fit(X_train_3c, y_train_3c) #correlation model perfomance
pipe_variance.fit(X_train_3c) #correlation variance
pipe_tree_importance.fit(X_train_3c,y_train_3c) #tree importance
pipe_model_performance_tree.fit(X_train_3c, y_train_3c)  #tree + model perfomance
pipe_model_variance_tree.fit(X_train_3c, y_train_3c) #tree + variance

#X_train
X_train_pefomance_3_pl = pipe_model_performance.transform(X_train_3c)
X_train_variance3_pl= pipe_variance.transform(X_train_3c)
X_train_tree3_pl= pipe_tree_importance.transform(X_train_3c)
X_train_performance_tree3_pl= pipe_model_performance_tree.transform(X_train_3c)
X_train_variance_tree3_pl= pipe_model_variance_tree.transform(X_train_3c)

#X_test
X_test_pefomance_3_pl = pipe_model_performance.transform(X_test_3c) #correlation model perfomance
X_test_variance3_pl= pipe_variance.transform(X_test_3c) #correlation variance
X_test_tree3_pl=pipe_tree_importance.transform(X_test_3c) #tree importance
X_test_performance_tree3_pl=pipe_model_performance_tree.transform(X_test_3c) #tree + model perfomance
X_test_variance_tree3_pl=pipe_model_variance_tree.transform(X_test_3c)#tree + variance



In [ ]:
print('X_train', X_train_pefomance_3_pl.shape, X_train_variance3_pl.shape, X_train_tree3_pl.shape, X_train_performance_tree3_pl.shape,X_train_variance_tree3_pl.shape,  '\n',
      'X_test',X_test_pefomance_3_pl.shape, X_test_variance3_pl.shape,X_test_tree3_pl.shape, X_test_performance_tree3_pl.shape, X_test_variance_tree3_pl.shape )

X_train (314327, 23) (314327, 23) (314327, 22) (314327, 11) (314327, 7) 
 X_test (69183, 23) (69183, 23) (69183, 22) (69183, 11) (69183, 7)


## 2 Models peformance

In [ ]:
#Function to evaluate the perfomance

classifiers = {
    'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42),
    'Extra Trees': ExtraTreesClassifier(n_estimators=100, random_state=42),
    'XGBoost': XGBClassifier()
}
def evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers):
    results = {}
    for name, clf in classifiers.items():
        clf.fit(X_train_dataset, y_train_dataset)

        # Train set predictions
        train_preds = clf.predict(X_train_dataset)
        train_f1 = f1_score(y_train_dataset, train_preds, average='macro')

        # Test set predictions
        test_preds = clf.predict(X_test_dataset)
        test_f1 = f1_score(y_test_dataset, test_preds, average='macro')

        results[name] = {'Train F1 Macro': train_f1, 'Test F1 Macro': test_f1}
    return results



In [ ]:
#Iterate over each dataset and evaluate

datasets = {
    'Performance 2': (X_train_pefomance_2_pl, X_test_pefomance_2_pl, y_train_2c, y_test_2),
    'Variance 2': (X_train_variance2_pl, X_test_variance2_pl,  y_train_2c, y_test_2),
    'Tree 2': (X_train_tree2_pl, X_test_tree2_pl,  y_train_2c, y_test_2),
    'Tree_perfomance 2': (X_train_performance_tree2_pl, X_test_performance_tree2_pl,  y_train_2c, y_test_2),
    'Tree_variance 2': (X_train_variance_tree2_pl, X_test_variance_tree2_pl,  y_train_2c, y_test_2),

    'Performance 2a': (X_train_pefomance_2a_pl, X_test_pefomance_2a_pl, y_train_2ac, y_test_2a),
    'Variance 2a': (X_train_variance2a_pl, X_test_variance2a_pl,y_train_2ac, y_test_2a),
    'Tree 2a': (X_train_tree2a_pl, X_test_tree2a_pl,y_train_2ac, y_test_2a),
    'Tree_perfomance 2a': (X_train_performance_tree2a_pl, X_test_performance_tree2a_pl,y_train_2ac, y_test_2a),
    'Tree_variance 2a': (X_train_variance_tree2a_pl, X_test_variance_tree2a_pl,y_train_2ac, y_test_2a),

    'Performance 3': (X_train_pefomance_3_pl, X_test_pefomance_3_pl,y_train_3c, y_test_3),
    'Variance 3': (X_train_variance3_pl, X_test_variance3_pl,y_train_3c, y_test_3),
    'Tree 3': (X_train_tree3_pl, X_test_tree3_pl,y_train_3c, y_test_3),
    'Tree_perfomance 3': (X_train_performance_tree3_pl, X_test_performance_tree3_pl,y_train_3c, y_test_3),
    'Tree_variance 3': (X_train_variance_tree3_pl, X_test_variance_tree3_pl,y_train_3c, y_test_3)
    }



final_results = {}
for dataset_name, (X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset) in datasets.items():
    print(f"Evaluating {dataset_name}...")
    results = evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers)
    final_results[dataset_name] = results


Evaluating Performance 2...
Evaluating Variance 2...
Evaluating Tree 2...
Evaluating Tree_perfomance 2...
Evaluating Tree_variance 2...
Evaluating Performance 2a...
Evaluating Variance 2a...
Evaluating Tree 2a...
Evaluating Tree_perfomance 2a...
Evaluating Tree_variance 2a...
Evaluating Performance 3...
Evaluating Variance 3...
Evaluating Tree 3...
Evaluating Tree_perfomance 3...
Evaluating Tree_variance 3...


In [ ]:
# Print the results
for dataset_name, classifiers_results in final_results.items():
    print(f"\nResults for {dataset_name}:")
    for clf_name, scores in classifiers_results.items():
        print(f"{clf_name} - Train F1 Macro: {scores['Train F1 Macro']:.4f}, Test F1 Macro: {scores['Test F1 Macro']:.4f}")



Results for Performance 2:
Random Forest - Train F1 Macro: 0.9995, Test F1 Macro: 0.6914
Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.7067
XGBoost - Train F1 Macro: 0.7962, Test F1 Macro: 0.6009

Results for Variance 2:
Random Forest - Train F1 Macro: 0.9995, Test F1 Macro: 0.6898
Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.6839
XGBoost - Train F1 Macro: 0.7949, Test F1 Macro: 0.6008

Results for Tree 2:
Random Forest - Train F1 Macro: 0.9988, Test F1 Macro: 0.6161
Extra Trees - Train F1 Macro: 0.9988, Test F1 Macro: 0.6233
XGBoost - Train F1 Macro: 0.7503, Test F1 Macro: 0.5657

Results for Tree_perfomance 2:
Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6236
Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.6543
XGBoost - Train F1 Macro: 0.7598, Test F1 Macro: 0.5478

Results for Tree_variance 2:
Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6213
Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.6516
XGBoost - Train F1 Macro: 0.75

Observations based on the results:

Overfitting: Both Random Forest and Extra Trees classifiers show signs of overfitting, as indicated by the near-perfect F1 scores on the training data and significantly lower scores on the test data. XGBoost, on the other hand, seems to generalize better to the test data, as the gap between its training and test scores is smaller.


Best Model: While XGBoost has a lower tendency to overfit, its test performance is not always the highest among the three classifiers. However, given its better generalization, it might be a safer choice in a real-world scenario where the data distribution can change.


Feature Selection: The results vary based on the feature selection method used (Performance, Variance, Tree). It's essential to understand the nature of these feature selection methods and choose the one that makes the most sense for your specific problem.





### 2.1 Best results


**Results for Performance 2:**

Random Forest - Train F1 Macro: 0.9995, Test F1 Macro: 0.6914

Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.7067

XGBoost - Train F1 Macro: 0.7962, Test F1 Macro: 0.6009

**Results for Variance 2a:**

Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6917

Extra Trees - Train F1 Macro: 0.9994, Test F1 Macro: 0.6870

XGBoost - Train F1 Macro: 0.7906, Test F1 Macro: 0.5980


**Results for Performance 2a:**

Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6867

Extra Trees - Train F1 Macro: 0.9994, Test F1 Macro: 0.6850

XGBoost - Train F1 Macro: 0.7904, Test F1 Macro: 0.5951


## 4 API

In [ ]:
#Converting to dataframe
X_train_perfomance_2_pl_df = pd.DataFrame(X_train_pefomance_2_pl)
X_train_variance2a_pl_df = pd.DataFrame(X_train_variance2a_pl)
X_train_perfomance_2a_pl_df = pd.DataFrame(X_train_pefomance_2a_pl)

X_test_perfomance_2_pl_df = pd.DataFrame(X_test_pefomance_2_pl)
X_test_variance2a_pl_df = pd.DataFrame(X_test_variance2_pl)
X_test_perfomance_2a_pl_df = pd.DataFrame(X_test_pefomance_2a_pl)


In [ ]:
# Concatenating the best perfomnces
data_train_perfomance_2 = pd.concat([X_train_pefomance_2_pl_df, y_train_2c], axis=1)
data_train_perfomance_2a = pd.concat([X_train_pefomance_2a_pl_df, y_train_2ac], axis=1)
data_train_variance_2a = pd.concat([X_train_variance2a_pl_df, y_train_2ac], axis=1)

In [ ]:
# Api for perfomance_2
clf = setup(data=data_train_perfomance_2, target='mode_main', session_id=0)

In [ ]:
best_models = compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

In [ ]:
best_models

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=0, verbose=0, warm_start=False)

In [ ]:
# Api for perfomance_2
clf = setup(data=data_train_perfomance_2a, target='mode_main', session_id=0)

In [ ]:
best_models = compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

In [ ]:
best_models

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=0, verbose=0, warm_start=False)

In [ ]:
# Api for perfomance_2
clf = setup(data=data_train_variance_2a, target='mode_main', session_id=0)

In [ ]:
best_models = compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

In [ ]:
best_models

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=0, verbose=0, warm_start=False)

After applying the API we found out that there is 2 new models that have not been considerated.  	**K Neighbors Classifier and Decision Tree Classifier**. Let's see how is the train vs test behavior of the selected dataset with these models.

### 4.1 K Neighbours Classifier and Decision Tree Classifier

In [ ]:
#Function to evaluate the perfomance

classifiers = {
     'K Neighbors': KNeighborsClassifier(n_neighbors=5),
    'Decision Tree': DecisionTreeClassifier(random_state=0)
}
def evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers):
    results = {}
    for name, clf in classifiers.items():
        clf.fit(X_train_dataset, y_train_dataset)

        # Train set predictions
        train_preds = clf.predict(X_train_dataset)
        train_f1 = f1_score(y_train_dataset, train_preds, average='macro')

        # Test set predictions
        test_preds = clf.predict(X_test_dataset)
        test_f1 = f1_score(y_test_dataset, test_preds, average='macro')

        results[name] = {'Train F1 Macro': train_f1, 'Test F1 Macro': test_f1}
    return results


In [ ]:

datasets = {
    'Performance 2': (X_train_pefomance_2_pl, X_test_pefomance_2_pl, y_train_2c, y_test_2),
    'Performance 2a': (X_train_pefomance_2a_pl, X_test_pefomance_2a_pl, y_train_2ac, y_test_2a),
    'Variance 2a': (X_train_variance2a_pl, X_test_variance2a_pl,y_train_2ac, y_test_2a),

    }

final_results = {}
for dataset_name, (X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset) in datasets.items():
    print(f"Evaluating {dataset_name}...")
    results = evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers)
    final_results[dataset_name] = results


In [ ]:
#Function to evaluate the perfomance
classifiers = {
    'K Neighbors': KNeighborsClassifier(n_neighbors=5),
    'Decision Tree': DecisionTreeClassifier(random_state=0)
}

def evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers):
    results = {}
    for name, clf in classifiers.items():

        # Apply Min-Max scaling if the classifier is K Neighbors
        if isinstance(clf, KNeighborsClassifier):
            scaler = MinMaxScaler()
            X_train_dataset = scaler.fit_transform(X_train_dataset)
            X_test_dataset = scaler.transform(X_test_dataset)

        clf.fit(X_train_dataset, y_train_dataset)

        # Train set predictions
        train_preds = clf.predict(X_train_dataset)
        train_f1 = f1_score(y_train_dataset, train_preds, average='macro')

        # Test set predictions
        test_preds = clf.predict(X_test_dataset)
        test_f1 = f1_score(y_test_dataset, test_preds, average='macro')

        results[name] = {'Train F1 Macro': train_f1, 'Test F1 Macro': test_f1}
    return results

datasets = {
    'Performance 2': (X_train_pefomance_2_pl, X_test_pefomance_2_pl, y_train_2c, y_test_2),
    'Performance 2a': (X_train_pefomance_2a_pl, X_test_pefomance_2a_pl, y_train_2ac, y_test_2a),
    'Variance 2a': (X_train_variance2a_pl, X_test_variance2a_pl,y_train_2ac, y_test_2a),
}

final_results = {}
for dataset_name, (X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset) in datasets.items():
    print(f"Evaluating {dataset_name}...")
    results = evaluate_performance(X_train_dataset, X_test_dataset, y_train_dataset, y_test_dataset, classifiers)
    final_results[dataset_name] = results


Evaluating Performance 2...
Evaluating Performance 2a...
Evaluating Variance 2a...


In [ ]:
# Print the results
for dataset_name, classifiers_results in final_results.items():
    print(f"\nResults for {dataset_name}:")
    for clf_name, scores in classifiers_results.items():
        print(f"{clf_name} - Train F1 Macro: {scores['Train F1 Macro']:.4f}, Test F1 Macro: {scores['Test F1 Macro']:.4f}")



Results for Performance 2:
K Neighbors - Train F1 Macro: 0.8971, Test F1 Macro: 0.5501
Decision Tree - Train F1 Macro: 0.9995, Test F1 Macro: 0.6031

Results for Performance 2a:
K Neighbors - Train F1 Macro: 0.8572, Test F1 Macro: 0.4670
Decision Tree - Train F1 Macro: 0.9994, Test F1 Macro: 0.5993

Results for Variance 2a:
K Neighbors - Train F1 Macro: 0.8565, Test F1 Macro: 0.4666
Decision Tree - Train F1 Macro: 0.9994, Test F1 Macro: 0.6061


## 5 Final Result


**Results for Performance 2:**

Random Forest - Train F1 Macro: 0.9995, Test F1 Macro: 0.6914

Extra Trees - Train F1 Macro: 0.9995, Test F1 Macro: 0.7067

**XGBoost - Train F1 Macro: 0.7962, Test F1 Macro: 0.6009**

**K Neighbors - Train F1 Macro: 0.8971, Test F1 Macro: 0.5501**

Decision Tree - Train F1 Macro: 0.9995, Test F1 Macro: 0.6031



**Results for Variance 2a:**

Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6917

Extra Trees - Train F1 Macro: 0.9994, Test F1 Macro: 0.6870

**XGBoost - Train F1 Macro: 0.7906, Test F1 Macro: 0.5980**

**K Neighbors - Train F1 Macro: 0.8565, Test F1 Macro: 0.4666**

Decision Tree - Train F1 Macro: 0.9994, Test F1 Macro: 0.6061



**Results for Performance 2a:**

Random Forest - Train F1 Macro: 0.9994, Test F1 Macro: 0.6867

Extra Trees - Train F1 Macro: 0.9994, Test F1 Macro: 0.6850

**XGBoost - Train F1 Macro: 0.7904, Test F1 Macro: 0.5951**

**K Neighbors - Train F1 Macro: 0.8572, Test F1 Macro: 0.4670**

Decision Tree - Train F1 Macro: 0.9994, Test F1 Macro: 0.5993

Overfitting: **The Random Forest, Extra Trees, and Decision Tree classifiers** show signs of overfitting. Their training F1 scores are almost perfect (close to 1), but their test F1 scores are significantly lower. This indicates that these models have learned the training data very well but are not generalizing well to unseen data.


**XGBoost** Performance: The XGBoost classifier seems to be the most balanced in terms of the difference between training and testing F1 scores. It's not overfitting as much as the tree-based models. This makes it a good candidate for further tuning and optimization, especially given the imbalance in the data.


**K Neighbors Performance**: The K Neighbors classifier also doesn't show as much overfitting as the tree-based models, but its performance is lower than XGBoost. It might benefit from further hyperparameter tuning or from using a different distance metric, especially given the high dimensionality of the data.

So we are going to continued with the **XGBoost**

## 3 Save the data

In [ ]:
#  Dataframes perfomance dataset 2
X_train_final_p2 = pd.DataFrame(X_train_perfomance_2_pl_df, columns=X_train_perfomance_2_pl_df.columns)
X_test_final_p2 = pd.DataFrame(X_test_perfomance_2_pl_df, columns=X_test_perfomance_2_pl_df.columns)
y_train_final_p2 = pd.DataFrame(y_train_2c)
y_test_final_p2 = pd.DataFrame(y_test_2c)


In [ ]:
print(X_train_final_p2.shape,X_test_final_p2.shape, y_train_final_p2.shape, y_test_final_p2.shape )

(320978, 18) (69183, 18) (320978, 1) (69183, 1)


In [ ]:
#  Dataframes perfomance dataset 2a
X_train_final_p2a = pd.DataFrame(X_train_perfomance_2a_pl_df, columns=X_train_perfomance_2a_pl_df.columns)
X_test_final_p2a = pd.DataFrame(X_test_perfomance_2a_pl_df, columns=X_test_perfomance_2a_pl_df.columns)
y_train_final_p2a = pd.DataFrame(y_train_2ac)
y_test_final_p2a = pd.DataFrame(y_test_2ac)

In [ ]:
print(X_train_final_p2a.shape,X_test_final_p2a.shape, y_train_final_p2a.shape, y_test_final_p2a.shape )

(322474, 19) (69183, 19) (322474, 1) (69183, 1)


In [ ]:
#  Dataframes variance dataset 2a
X_train_var_p2a = pd.DataFrame(X_train_variance2a_pl_df, columns=X_train_variance2a_pl_df.columns)
X_test_var_p2a = pd.DataFrame(X_test_variance2a_pl_df, columns=X_test_variance2a_pl_df.columns)
y_train_var_p2a = pd.DataFrame(y_train_2ac)
y_test_var_p2a = pd.DataFrame(y_test_2ac)



In [ ]:
print(X_train_var_p2a.shape,X_test_var_p2a.shape, y_train_var_p2a.shape, y_test_var_p2a.shape )

(322474, 19) (69183, 18) (322474, 1) (69183, 1)


In [ ]:
# Columns in X_train_final_p2a but not in the other two
unique_final_p2a = set(X_train_final_p2a.columns) - set(X_train_var_p2a.columns) - set(X_train_final_p2.columns)

# Columns in X_train_var_p2a but not in the other two
unique_var_p2a = set(X_train_var_p2a.columns) - set(X_train_final_p2a.columns) - set(X_train_final_p2.columns)

# Columns in X_train_final_p2 but not in the other two
unique_final_p2 = set(X_train_final_p2.columns) - set(X_train_final_p2a.columns) - set(X_train_var_p2a.columns)

print("Unique columns in X_train_final_p2a:", unique_final_p2a)
print("Unique columns in X_train_var_p2a:", unique_var_p2a)
print("Unique columns in X_train_final_p2:", unique_final_p2)


Unique columns in X_train_final_p2a: {'bicycles_arb_disc', 'green_tr', 'diversity_tr'}
Unique columns in X_train_var_p2a: {'age', 'diversity_arb_disc'}
Unique columns in X_train_final_p2: {'temp_tr', 'distance_tr', 'diversity', 'wind_tr'}


In [ ]:
data_dir = ('/Users/aprilm/Desktop/TFG/data_fs/')

# Save dataframes baseline to CSV files in the specified directory
#  Dataframes perfomance dataset 2

X_train_final_p2.to_csv(data_dir + 'X_train_p2.csv', index=False)
X_test_final_p2.to_csv(data_dir + 'X_test_p2.csv', index=False)
y_train_final_p2.to_csv(data_dir + 'y_train_p2.csv', index=False)
y_test_final_p2.to_csv(data_dir + 'y_test_p2.csv', index=False)


In [ ]:

# Save dataframes baseline to CSV files in the specified directory
#  Dataframes perfomance dataset 2a

X_train_final_p2a.to_csv(data_dir + 'X_train_p2a.csv', index=False)
X_test_final_p2a.to_csv(data_dir + 'X_test_p2a.csv', index=False)
y_train_final_p2a.to_csv(data_dir + 'y_train_p2a.csv', index=False)
y_test_final_p2a.to_csv(data_dir + 'y_test_p2a.csv', index=False)

In [ ]:

# Save dataframes baseline to CSV files in the specified directory
#  Dataframes variance dataset 2a

X_train_var_p2a.to_csv(data_dir + 'X_train_v2a.csv', index=False)
X_test_var_p2a.to_csv(data_dir + 'X_test_v2a.csv', index=False)
y_train_var_p2a.to_csv(data_dir + 'y_train_v2a.csv', index=False)
y_test_var_p2a.to_csv(data_dir + 'y_test_v2a.csv', index=False)